### PR TITLE
Add localfile wildcard support for windows IT

### PR DESCRIPTION
**Pull request by Deblintrake09**
|Related issue|
|-------------|
| #3786     |

## Description

This PR adds test support for wildcards usage in the localfile `location` tag for Windows agents.

### Added

- Added 6 test cases in `/test_logcollector/test_location_wildcards` suite. Testing different wildcards combinations and a case where wildcards do not match a valid file

## Testing performed

| Tester             | Test path | Jenkins | Local  | OS | Commit | Notes                |
|--------------------|-----------|---------|--------|-----|--------|----------------------|
| @Deblintrake09  (Developer)  | test_logcollector           |   | [:green_circle::green_circle::green_circle:](https://github.com/wazuh/wazuh-qa/files/11056100/Manager_Results.zip)  |  Manager        |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/35f4ac6330db8b957231daa9f2131cf30eeaab0c      | Nothing to highlight |
| @Deblintrake09  (Developer)  | test_logcollector           |   | [:green_circle::green_circle::green_circle:](https://github.com/wazuh/wazuh-qa/files/11055296/LinuxAgent_Results.zip) |  Linux agent        |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/35f4ac6330db8b957231daa9f2131cf30eeaab0c      | Nothing to highlight |
| @Deblintrake09  (Developer)  | test_logcollector           |   | [:green_circle::green_circle::green_circle:](https://github.com/wazuh/wazuh-qa/files/11055096/WinAgent_Results.zip)  |  Windows agent        |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/35f4ac6330db8b957231daa9f2131cf30eeaab0c      | Nothing to highlight |
| @damarisg  (Reviewer)   | test_logcollector   | [🟢](https://ci.wazuh.info/job/Test_integration/38028/)  | :no_entry_sign: :no_entry_sign: :no_entry_sign:  |Manager         |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/ae13d712687903ffaf77a53ca25e689293f45e8f | Nothing to highlight |
| @damarisg  (Reviewer)   | test_logcollector    | [🟢](https://ci.wazuh.info/job/Test_integration/38029/)  | :no_entry_sign: :no_entry_sign: :no_entry_sign:  | Linux agent         |  https://github.com/wazuh/wazuh-qa/pull/3906/commits/ae13d712687903ffaf77a53ca25e689293f45e8f | Nothing to highlight |
| @damarisg  (Reviewer)   | test_logcollector    | [🟢](https://ci.wazuh.info/job/Test_integration/38030/)  | :no_entry_sign: :no_entry_sign: :no_entry_sign:  |  Windows agent | https://github.com/wazuh/wazuh-qa/pull/3906/commits/ae13d712687903ffaf77a53ca25e689293f45e8f      | Nothing to highlight |
| @damarisg  (Reviewer)   | test_logcollector    | [🟢](https://ci.wazuh.info/view/Tests/job/Test_integration/38062/)  | :no_entry_sign: :no_entry_sign: :no_entry_sign:  | Solaris and macOS agent         |  https://github.com/wazuh/wazuh-qa/pull/3906/commits/ae13d712687903ffaf77a53ca25e689293f45e8f | Nothing to highlight |
| @Deblintrake09  (Developer)  | test_logcollector           |   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/11402269/Manager.zip)  |  Manager        |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/5de9cca5c1d333098c942e9902d8aab8ec17f840      | Additional results after moving test to new folder |
| @Deblintrake09  (Developer)  | test_logcollector           |   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/11402268/LinuxAgent.zip) |  Linux agent        |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/5de9cca5c1d333098c942e9902d8aab8ec17f840      | Additional results after moving test to new folder |
| @Deblintrake09  (Developer)  | test_logcollector           |   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/11402270/WinAgent.zip)  |  Windows agent        |   https://github.com/wazuh/wazuh-qa/pull/3906/commits/5de9cca5c1d333098c942e9902d8aab8ec17f840      | Additional results after moving test to new folder |

### CODE REVIEW

File: deps/wazuh_testing/wazuh_testing/tools/file.py [hunk 1]
```diff
@@ -236,7 +236,7 @@ def delete_file(file_path):
 
 def delete_path_recursively(path):
     if os.path.exists(path):
-        shutil.rmtree(path, onerror=on_write_error)
+        shutil.rmtree(path, ignore_errors=True, onerror=on_write_error)
```
Comment (damarisg) [inline deps/wazuh_testing/wazuh_testing/tools/file.py:239]: Why do you ignore the errors?

File: tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
+'''
+copyright: Copyright (C) 2015-2022, Wazuh Inc.
```
Comment (damarisg) [inline tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py:2]:
```suggestion
copyright: Copyright (C) 2015-2023, Wazuh Inc.
```

File: tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        - r".*wazuh-agent.*expand_win32_wildcards.*DEBUG: No file/folder that matches {regex}"
+        - r".*wazuh-agent.*check_pattern_expand.*New file that matches the '{file_path}' pattern: '(.*)'"
+        - r".*wazuh-agent.*Analizing file: '{file}'.*"
+        - r".*wazuh-agent.*DEBUG: Reading syslog '{message}'.*"
+    '''
+    command = f"echo {log_sample}>> {test_file}"
+    file = re.escape(test_file)
+
+    if not metadata['matches']:
+        # If it does not match, check that mesage shows no matching file was found
```
Comment (damarisg) [inline tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py:154]:
```suggestion
        # If it does not match, check that message shows no matching file was found
```

**Review by damarisg — CHANGES_REQUESTED** [review]

File: tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py (hunk 3 above)
Comment (Deblintrake09) [reply tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py]: Fixed in https://github.com/wazuh/wazuh-qa/pull/3906/commits/35f4ac6330db8b957231daa9f2131cf30eeaab0c

File: tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py (hunk 2 above)
Comment (Deblintrake09) [reply tests/integration/test_logcollector/test_location_wildcards/test_location_wildcards.py]: Fixed in https://github.com/wazuh/wazuh-qa/pull/3906/commits/35f4ac6330db8b957231daa9f2131cf30eeaab0c

File: deps/wazuh_testing/wazuh_testing/tools/file.py (hunk 1 above)
Comment (Deblintrake09) [reply deps/wazuh_testing/wazuh_testing/tools/file.py]: Fixed in https://github.com/wazuh/wazuh-qa/pull/3906/commits/35f4ac6330db8b957231daa9f2131cf30eeaab0c

**Review by damarisg — CHANGES_REQUESTED** [review]
I need that you check the executions, because I found [failures](https://ci.wazuh.info/view/Tests/job/Test_integration/37110/artifact/windows_agent_tests_Test_integration_B37110_20230330185059.txt).

**Review by damarisg — APPROVED** [review]

File: deps/wazuh_testing/wazuh_testing/modules/logcollector/event_monitor.py [hunk 4]
```diff
@@ -132,3 +133,43 @@ def check_ignore_restrict_message_not_found(message, regex, tag, prefix):
...
+
+    Returns: True if the expected message has been found, False otherwise.
+    """
+    callback_msg = f".*New file that matches the '{location_regex}' pattern: '{file_path}'"
+
+    return check_logcollector_event(file_monitor=file_monitor, timeout=timeout, callback=callback_msg,
+                                    error_message=error_message, prefix=prefix, escape=escape)
+
+
+def check_wildcard_pattern_no_match(regex, prefix, error_message=None, file_monitor=None, timeout=T_10, escape=False):
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/modules/logcollector/event_monitor.py:159]:
```suggestion
def check_win_wildcard_pattern_no_match(regex, prefix, error_message=None, file_monitor=None, timeout=T_10, escape=False):
```

It seems that this log only applies to Windows.

File: tests/integration/test_logcollector/test_configuration/test_basic_configuration_log_format.py [hunk 5]
```diff
@@ -274,7 +274,7 @@ def check_log_format_invalid(cfg):
                             error_message=gc.GENERIC_CALLBACK_ERROR_MESSAGE)
 
     if sys.platform != 'win32':
-        log_callback = gc.callback_error_in_configuration('ERROR', prefix,
+        log_callback = gc.callback_error_in_configuration('CRITICAL', prefix,
```
Comment (juliamagan) [inline tests/integration/test_logcollector/test_configuration/test_basic_configuration_log_format.py:277]: Check this change: https://github.com/wazuh/wazuh-qa/pull/4094

File: tests/integration/test_logcollector/test_location_wildcards/data/test_cases/cases_location_wildcards.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
+- name: Test single asterisk wildcard
```
Comment (juliamagan) [inline tests/integration/test_logcollector/test_location_wildcards/data/test_cases/cases_location_wildcards.yaml:1]: We have to use names compatible with the `-k` parameter.

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ!

Since these tests only apply to Windows, I would rename them by adding `win` somewhere for clarification, and add them to the `test_location` module, instead of creating a new one, since they test the same thing.

File: deps/wazuh_testing/wazuh_testing/modules/logcollector/event_monitor.py (hunk 4 above)
Comment (Deblintrake09) [reply deps/wazuh_testing/wazuh_testing/modules/logcollector/event_monitor.py]: Done in https://github.com/wazuh/wazuh-qa/pull/3906/commits/ffca56f7b9804b4ef1df2c7ffdd10987414e6e48

File: tests/integration/test_logcollector/test_location_wildcards/data/test_cases/cases_location_wildcards.yaml (hunk 6 above)
Comment (Deblintrake09) [reply tests/integration/test_logcollector/test_location_wildcards/data/test_cases/cases_location_wildcards.yaml]: Done in https://github.com/wazuh/wazuh-qa/pull/3906/commits/ffca56f7b9804b4ef1df2c7ffdd10987414e6e48

File: tests/integration/test_logcollector/test_configuration/test_basic_configuration_log_format.py (hunk 5 above)
Comment (Deblintrake09) [reply tests/integration/test_logcollector/test_configuration/test_basic_configuration_log_format.py:277]: This was applied after Jenkins instances for this Issue failed on this tests. It was also failing on local executions.

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,7 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 
+- Add Windows location wildcards tests ([#3906](https://github.com/wazuh/wazuh-qa/pull/3906)) \- (Tests)
```
Comment (juliamagan) [inline CHANGELOG.md:9]:
```suggestion
- Add Windows location wildcards tests ([#3906](https://github.com/wazuh/wazuh-qa/pull/3906)) \- (Tests + Framework)
```

**Review by juliamagan — CHANGES_REQUESTED** [review]

**Review by juliamagan — APPROVED** [review]
LGTM

**Comment by Deblintrake09** [comment]
Branch was recreated in #4263 since this branch has unverified commits.

**Pull request closed without merging**